python/ray/data/_internal/stats.py

```diff
@@ -275,137 +332,302 @@ def summary_string(
+            out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         return out
 
-    def _summarize_iter(self) -> str:
-        out = ""
-        if (
-            self.iter_total_s.get()
-            or self.iter_wait_s.get()
-            or self.iter_next_batch_s.get()
-            or self.iter_format_batch_s.get()
-            or self.iter_get_s.get()
-        ):
-            out += "\nDataset iterator time breakdown:\n"
-            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s.get()))
-            out += "* In ray.get(): {}\n".format(fmt(self.iter_get_s.get()))
-            out += "* In next_batch(): {}\n".format(fmt(self.iter_next_batch_s.get()))
-            out += "* In format_batch(): {}\n".format(
-                fmt(self.iter_format_batch_s.get())
-            )
-            out += "* In user code: {}\n".format(fmt(self.iter_user_s.get()))
-            out += "* Total time: {}\n".format(fmt(self.iter_total_s.get()))
-        return out
+    def get_total_wall_time(self) -> float:
+        parent_sum = sum(p.get_total_wall_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_total_cpu_time(self) -> float:
+        parent_sum = sum(p.get_total_cpu_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.CPU_TIME).get("sum", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_max_heap_memory(self) -> float:
+        parent_memory = [p.get_max_heap_memory() for p in self.parents]
+        parent_max = max(parent_memory) if parent_memory else 0
+        return max(
+            parent_max,
+            *[
+                getattr(ss, StageStatsSummary.PEAK_HEAP_MEMORY).get("max", 0)
+                for ss in self.stages_stats
+            ],
+        )
+
+
+@dataclass
+class StageStatsSummary:
+    # Keys used in initialization params
+    BLOCK_EXECUTION_SUMMARY = "block_execution_summary_str"
+    WALL_TIME = "wall_time"
+    CPU_TIME = "cpu_time"
+    PEAK_HEAP_MEMORY = "memory"
+    OUTPUT_NUM_ROWS = "output_num_rows"
+    OUTPUT_SIZE_BYTES = "output_size_bytes"
+    NODE_COUNT = "node_count"
+
+    stage_name: str
+    is_substage: bool
+    time_total_s: float
+    block_execution_summary_str: str
+    wall_time: Optional[Dict[str, float]] = None
+    cpu_time: Optional[Dict[str, float]] = None
+    memory: Optional[Dict[str, float]] = None
+    output_num_rows: Optional[Dict[str, float]] = None
+    output_size_bytes: Optional[Dict[str, float]] = None
+    node_count: Optional[Dict[str, float]] = None
+
+    @classmethod
+    def from_block_metadata(
+        cls,
+        block_metas: List[BlockMetadata],
+        time_total_s: float,
+        stage_name: str,
+        is_substage: bool,
+    ) -> "StageStatsSummary":
+        """Calculate the stats for a given list of blocks from a stage,
```

Comment:
nit: I think it's "Calculate the stats for a stage from a given list of blocks", right?